include/unifex/sender_concepts.hpp

```diff
@@ -15,98 +15,310 @@
+#include <unifex/detail/fwddecls.hpp>
 #include <unifex/tag_invoke.hpp>
 #include <unifex/type_traits.hpp>
+#include <unifex/receiver_concepts.hpp>
 
+#include <exception>
 #include <tuple>
 #include <type_traits>
+#include <utility>
 
 #include <unifex/detail/prologue.hpp>
 
 namespace unifex {
-namespace _start {
-  inline constexpr struct _fn {
-   private:
-    template <bool>
-    struct _impl {
-      template <typename Operation>
-      auto operator()(Operation& op) const noexcept
-          -> tag_invoke_result_t<_fn, Operation&> {
-        static_assert(
-            is_nothrow_tag_invocable_v<_fn, Operation&>,
-            "start() customisation must be noexcept");
-        return unifex::tag_invoke(_fn{}, op);
-      }
-    };
-   public:
-      template <typename Operation>
-      auto operator()(Operation& op) const noexcept
-        -> callable_result_t<
-            _impl<is_tag_invocable_v<_fn, Operation&>>, Operation&> {
-      return _impl<is_tag_invocable_v<_fn, Operation&>>{}(op);
+
+struct sender_base {};
+
+template <typename>
+struct sender_traits;
+
+/// \cond
+namespace detail {
+  template <template <template <typename...> class, template <typename...> class> class>
+  struct _has_value_types;
+
+  template <template <template <typename...> class> class>
+  struct _has_error_types;
+
+  template <typename, typename>
+  struct _as_receiver;
+
+  template <typename R, typename E>
+  struct _as_invocable {
+    R* r_ ;
+    explicit _as_invocable(R& r) noexcept
+      : r_(std::addressof(r)) {}
+    _as_invocable(_as_invocable&& other) noexcept
+      : r_(std::exchange(other.r_, nullptr)) {}
+    ~_as_invocable() {
+      if(r_)
+        unifex::set_done((R&&) *r_);
     }
-  } start{};
+    void operator()() & noexcept try {
+      unifex::set_value((R&&) *r_);
+      r_ = nullptr;
+    } catch(...) {
+      unifex::set_error((R&&) *r_, std::current_exception());
+      r_ = nullptr;
+    }
+  };
+
+  template <typename S>
+  UNIFEX_CONCEPT_FRAGMENT(  //
+    _has_sender_types_impl, //
+      requires() (          //
+        // BUGBUG TODO:
+        // typename (std::bool_constant<S::sends_done>),
+        typename (_has_value_types<S::template value_types>),
+        typename (_has_error_types<S::template error_types>)
+      ));
+  template <typename S>
+  UNIFEX_CONCEPT        //
+    _has_sender_types = //
+      UNIFEX_FRAGMENT(detail::_has_sender_types_impl, S);
+
+  template <typename S>
+  UNIFEX_CONCEPT_FRAGMENT(  //
+    _not_has_sender_traits, //
+      requires() (          //
+        typename (typename sender_traits<S>::_unspecialized)
+      ));
+  template <typename S>
+  UNIFEX_CONCEPT         //
+    _has_sender_traits = //
+      !UNIFEX_FRAGMENT(detail::_not_has_sender_traits, S);
 
-  template <>
-  struct _fn::_impl<false> {
-    template <typename Operation>
+  struct _void_sender_traits {
+    template <template <typename...> class Tuple, template <typename...> class Variant>
```

Comment:
The order of the args here is the reverse of the order used everywhere in libunifex.
ie. senders in libunifex have the value_types alias take the Variant arg first followed by the Tuple arg.

Reply:
It's backwards of how P0443 does it. But I agree this PR should be consistent with the rest of libunifex, and if we change it we do it all once, and consistently.